### ISSUE TITLE
How to get notifier in json response?

### ISSUE THREAD

**Issue by walidvb** [simukappu/activity_notification#108]
Hi,

First, thanks for the great job on this gem.
Second, apols if this is rather a stackoverflow question. Please let me know if I should rather post it there.

Now, I'm having doubts on how to implement some behaviours – along with doubts on some choices, and therefore on whether I am using the gem correctly.

Q1) Why would the json api return the `target`(supposedly the person reading the notification, right?)
S2) To retrieve the `notifier_avatar_url`, one would need to override the `notifications_controller` as below?
S3) Is there any built-in way to retrieve only the most recent notifications?
Q4) To add some metadata to a notification, I've created a CustomNotification model, that has a `metadata` column, which notifies its `target`. This allows me to pass any sort of data, but I lose the built-in link to the real `notifiable`. Is there any better way to handle this?

I feel like i'm going through too many hoops to do smth that seems to me fairly standard?

Thanks,

Walid 

S2)
```
  def index
    set_index_options
    load_index if params[:reload].to_s.to_boolean(true)
    respond_to do |format|
      format.html # index.html.erb
      format.json { render json: @notifications.to_json(include: [:target, :notifiable, :group, notifier: {only: :first_name}]) }
    end
  end
```

S3)
```
  def set_index_options
    super
    if last_updated_at = params[:last_updated_at]
      @index_options[:custom_filter] = ['created_at > ?', last_updated_at]
    end
  end
```

**Comment by simukappu** [comment]
Hi,

Q1) You need to specify the target to get JSON response from notification controller. Notification model is associated target model and the controller just return Notification model as JSON format. There is no special reason why the json api return the target.

S2) You can include `notifier_avatar_url` column in your *Notifier* model. I don't think you have to customize the controller.

S3) Yes. You can specify the order of notification as default or [reverse](https://github.com/simukappu/activity_notification/blob/master/app/controllers/activity_notification/notifications_controller.rb#L15) param, default is latest first.
You can also specify the number of records in response as [limit](https://github.com/simukappu/activity_notification/blob/master/app/controllers/activity_notification/notifications_controller.rb#L14) param.
Therefore, when you request with `limit=1` and you will get only the most recent notification.

Q4) You should use *parameter* column of *Notification* model for that. You can set any parameter as Hash type to generated notification records. See [parameter option in notify method](https://github.com/simukappu/activity_notification/blob/master/lib/activity_notification/apis/notification_api.rb#L216) and related documents.

Please close this issue if your questions are resolved.

**Comment by simukappu** [comment]
Do you want *notifications_controller#index* JSON response include *notifier*?

```ruby
format.json { render json: @notifications.to_json(include: [:target, :notifiable, :group, *notifier]) }
```
https://github.com/simukappu/activity_notification/blob/master/app/controllers/activity_notification/notifications_controller.rb#L29

**Comment by walidvb** [comment]
Hi @simukappu, thanks for your reply.

Q1) Not sure I understand the comment, but i'm not having any issues on that side(specifically, JSON response is returned when the right headers are set)

Q2) great. I understand there is a list of columns automatically fetched from the `notifier` model? Is there a list anywhere? I'll try to take a look at the code, too

Q3) get it. However, this still doesn't give me the granularity i require. My use case is as follows: A device retrieves the latest notifications, and stores them in memory. On every subsequent request, it therefore only needs to receive the delta. Solution as per above works for me 👍🏻, but i thought this would be something fairly common? If so, i might create a PR to add this feature.

Q4) Great, I had missed that somehow. I think I still need to combine with my `CustomNotification`, as a notification needs to be attached to a model. FYI, that was unclear to me, and for the life of me i couldn't figure the magic out from the example app(i now opened the controller files, duh.)

Leaving this open for you to take a look, thanks for closing when you have / if appropriate!

Thanks for your time!

**Comment by simukappu** [comment]
Working in progress to create api mode for activity_notification.

**Comment by simukappu** [comment]
Released as v2.1.0. Thanks!

**Comment by walidvb** [comment]
Awesome, thanks a bunch! 
Will look at upgrading ASAP!
